les/clientpool.go

```diff
@@ -103,296 +106,180 @@ type clientPool struct {
 // clients have a limited access to IP addresses while new node keys can be easily
 // generated so it would be useless to assign a negative value to them.
 type clientPoolPeer interface {
-	ID() enode.ID
+	Node() *enode.Node
 	freeClientId() string
 	updateCapacity(uint64)
-	freezeClient()
+	freeze()
+	allowInactive() bool
 }
 
-// clientInfo represents a connected client
+// clientInfo defines all information required by clientpool.
 type clientInfo struct {
-	address                string
-	id                     enode.ID
-	connectedAt            mclock.AbsTime
-	capacity               uint64
-	priority               bool
-	pool                   *clientPool
-	peer                   clientPoolPeer
-	queueIndex             int // position in connectedQueue
-	balanceTracker         balanceTracker
-	posFactors, negFactors priceFactors
-	balanceMetaInfo        string
+	node        *enode.Node
+	address     string
+	peer        clientPoolPeer
+	connected   bool
+	connectedAt mclock.AbsTime
+	balance     *lps.NodeBalance
 }
 
-// connSetIndex callback updates clientInfo item index in connectedQueue
-func connSetIndex(a interface{}, index int) {
-	a.(*clientInfo).queueIndex = index
-}
-
-// connPriority callback returns actual priority of clientInfo item in connectedQueue
-func connPriority(a interface{}, now mclock.AbsTime) int64 {
-	c := a.(*clientInfo)
-	return c.balanceTracker.getPriority(now)
-}
-
-// connMaxPriority callback returns estimated maximum priority of clientInfo item in connectedQueue
-func connMaxPriority(a interface{}, until mclock.AbsTime) int64 {
-	c := a.(*clientInfo)
-	pri := c.balanceTracker.estimatedPriority(until, true)
-	c.balanceTracker.addCallback(balanceCallbackQueue, pri+1, func() {
-		c.pool.lock.Lock()
-		if c.queueIndex != -1 {
-			c.pool.connectedQueue.Update(c.queueIndex)
+// newClientPool creates a new client pool
+func newClientPool(lespayDb ethdb.Database, minCap, freeClientCap uint64, activeBias time.Duration, clock mclock.Clock, removePeer func(enode.ID)) *clientPool {
+	if minCap > freeClientCap {
+		panic(nil)
+	}
+	ns := nodestate.NewNodeStateMachine(nil, nil, clock, clientPoolSetup)
+	pool := &clientPool{
+		ns:                  ns,
+		BalanceTrackerSetup: balanceTrackerSetup,
+		PriorityPoolSetup:   priorityPoolSetup,
+		clock:               clock,
+		minCap:              minCap,
+		freeClientCap:       freeClientCap,
+		removePeer:          removePeer,
+	}
+	pool.ti = lps.NewTokenIssuer(ns, tokenIssuerSetup, clock, freeClientCap, activeBias/8, time.Hour)
+	pool.posExp = pool.ti.NewTokenExpirer()
+	pool.negExp = pool.ti.NewTokenExpirer()
```

Comment:
Now the token issuer is not used. Can we remove it now and add it in the next PR? the balancer and priority pool is complicated enough, would be easier by separating

Reply:
Fair enough, removed